src/glum/_transformers.py

```diff
@@ -0,0 +1,347 @@
+    CollinearityResults
+        Information about the columns to keep and the columns to drop.
+    """
+    if fit_intercept:
+        X = np.hstack((np.ones((X.shape[0], 1)), X))
+    if mode == "gram":
+        gram = X.T @ X
+        R, P = qr(gram, mode="r", pivoting=True)  # type: ignore
+    elif mode == "direct":
+        R, P = qr(X, mode="r", pivoting=True)  # type: ignore
+    else:
+        raise ValueError(f"Mode must be 'gram' or 'direct', got {mode}")
+        # Cannot happen
+
+    return _find_collinear_columns(R, P, fit_intercept, tolerance=tolerance)
+
+
+def _find_collinear_columns(
+    R: np.ndarray, P: np.ndarray, fit_intercept, tolerance: float = 1e-6
+) -> CollinearityResults:
+    """Determine the rank of X from the QR decomposition of X^T X.
+
+    Parameters
+    ----------
+    X : tm.MatrixBase
+        The design matrix.
+
+    Returns
+    -------
+    CollinearityResults
+        The indices of the columns to keep and the columns to drop.
+    """
+    permuted_keep_mask = np.abs(np.diag(R)) > tolerance
+    keep_mask = np.empty_like(permuted_keep_mask)
+    keep_mask[P] = permuted_keep_mask
+
+    keep_idx = np.where(keep_mask)[0]
+    drop_idx = np.where(~keep_mask)[0]
+
+    intercept_safe = False
+    if fit_intercept:
+        if 0 not in drop_idx:
+            intercept_safe = True
+        keep_idx -= 1
+        drop_idx -= 1
+
+    keep_idx = keep_idx[keep_idx != -1]
+    drop_idx = drop_idx[drop_idx != -1]
+
+    return CollinearityResults(keep_idx, drop_idx, intercept_safe)
+
+
+class ColumnMap(NamedTuple):
+    """Mapping from DataFrame to design matrix."""
+
+    column_pos: int
+    column_name: Hashable
+    categorical: bool
+    category: Optional[str] = None
+    base_category: Optional[str] = None
+
+
+def _get_column_mapping(X: pd.DataFrame) -> Sequence[ColumnMap]:
+    column_mapping = []
+    for column_pos, (column_name, dtype) in enumerate(X.dtypes.items()):
+        if isinstance(dtype, pd.CategoricalDtype):
+            if len(dtype.categories) > 1:
+                base_category = dtype.categories[0]
+                for category in dtype.categories[1:]:
+                    column_mapping.append(
+                        ColumnMap(
+                            column_pos, column_name, True, category, base_category
+                        )
+                    )
+        else:
+            column_mapping.append(ColumnMap(column_pos, column_name, False))
+    return column_mapping
+
+
+class Decollinearizer(TransformerMixin, BaseEstimator):
+    """Drop collinear columns from the design matrix implied by a dataset.
+
+    The type of the output is the same as the input. For non-categorical
+    columns, collinear ones are simply dropped. For categorical columns
+    (e.g. in a pandas.DataFrame or a tabmat.SplitMatrix), values whose
+    columns in the design matrix would be dropped are replaced with the
+    first category. This supposes that the first category is the reference,
```

Comment:
no Oxford comma with two items :)

Reply:
Which one? 😅 I admit that I'm a bit comma-happy in general but I can't find the problematic one here

Reply:
You are right, I misread here :).